### DISCUSSION TITLE
Renovate fails to maintain Gemfile.lock on Ruby version mismatch

### DISCUSSION THREAD

**Discussion by nfm** [renovatebot/renovate#10758] [Request Help]
**How are you running Renovate?**

- [ ] WhiteSource Renovate hosted app on github.com
- [x] Self hosted

If using the hosted app, please skip to the next section.
Otherwise, if self-hosted, please complete the following:

Please select which platform you are using:

- [ ] Azure DevOps (dev.azure.com)
- [ ] Azure DevOps Server
- [ ] Bitbucket Cloud (bitbucket.org)
- [ ] Bitbucket Server
- [ ] Gitea
- [ ] github.com
- [ ] GitHub Enterprise Server
- [ ] gitlab.com
- [x] GitLab self-hosted

Renovate version: 25.52.7

**Describe the bug**

Renovate opens MRs for rubygem updates and modifies the version in our `Gemfile`, but doesn't update the `Gemfile.lock` when it should.

From the renovate logs, it looks like it's hitting an error with mismatching ruby versions.

Our `Gemfile` does specify a ruby version (currently `ruby "2.7.3"`).

**Relevant debug logs**

<details><summary>Click me to see logs</summary>

```
 INFO: Gemfile.lock update failed due to an unknown reason (repository=**redacted**, branch=renovate/newrelic_rpm-7.x)
       "err": {
         "killed": false,
         "code": 7,
         "signal": null,
         "cmd": "bundle lock --update newrelic_rpm",
         "stdout": "Fetching gem metadata from https://rubygems.org/.........\n",
         "stderr": "Could not find gem 'Ruby\u0000 (~> 2.7.3.0)' in the local ruby installation.\nThe source contains the following versions of 'Ruby\u0000': 3.0.1.64\n",
         "message": "Command failed: bundle lock --update newrelic_rpm\nCould not find gem 'Ruby\u0000 (~> 2.7.3.0)' in the local ruby installation.\nThe source contains the following versions of 'Ruby\u0000': 3.0.1.64\n",
         "stack": "Error: Command failed: bundle lock --update newrelic_rpm\nCould not find gem 'Ruby\u0000 (~> 2.7.3.0)' in the local ruby installation.\nThe source contains the following versions of 'Ruby\u0000': 3.0.1.64\n\n    at ChildProcess.exithandler (child_process.js:319:12)\n    at ChildProcess.emit (events.js:376:20)\n    at ChildProcess.emit (domain.js:470:12)\n    at maybeClose (internal/child_process.js:1055:16)\n    at Process.ChildProcess._handle.onexit (internal/child_process.js:288:5)"
       }
 INFO: Branch created (repository=**redacted**, branch=renovate/newrelic_rpm-7.x)
       "commitSha": "7df89dd"
 INFO: PR created (repository=**redacted**, branch=renovate/newrelic_rpm-7.x)
       "pr": 6056,
       "prTitle": "Update dependency newrelic_rpm to ~>7.2.0"
 WARN: artifactErrors (repository=**redacted**, branch=renovate/newrelic_rpm-7.x)
       "artifactErrors": [
         {
           "lockFile": "Gemfile.lock",
           "stderr": "Fetching gem metadata from https://rubygems.org/.........\n\nCould not find gem 'Ruby\u0000 (~> 2.7.3.0)' in the local ruby installation.\nThe source contains the following versions of 'Ruby\u0000': 3.0.1.64\n"
         }
       ]
ERROR: Error ensuring PR: Response code 400 (Bad Request) (repository=**redacted**, branch=renovate/newrelic_rpm-7.x)
       "err": {
         "name": "HTTPError",
         "timings": {
           "start": 1625573925144,
           "socket": 1625573925144,
           "lookup": 1625573925145,
           "connect": 1625573925150,
           "secureConnect": 1625573925154,
           "upload": 1625573925154,
           "response": 1625573925242,
           "end": 1625573925243,
           "phases": {
             "wait": 0,
             "dns": 1,
             "tcp": 5,
             "tls": 4,
             "request": 0,
             "firstByte": 88,
             "download": 1,
             "total": 99
           }
         },
         "message": "Response code 400 (Bad Request)",
         "stack": "HTTPError: Response code 400 (Bad Request)\n    at Request.<anonymous> (/usr/src/app/node_modules/got/dist/source/as-promise/index.js:117:42)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)",
         "options": {
           "headers": {
             "user-agent": "https://github.com/renovatebot/renovate",
             "private-token": "**redacted**",
             "accept": "application/json",
             "content-type": "application/json",
             "content-length": "818",
             "accept-encoding": "gzip, deflate, br"
           },
           "url": "https://gitlab.com/api/v4/projects/**redacted**/merge_requests/**redacted**/notes",
           "username": "",
           "password": "",
           "method": "POST",
           "http2": false
         },
         "response": {
           "statusCode": 400,
           "statusMessage": "Bad Request",
           "body": "Bad Request",
           "headers": {
             "date": "Tue, 06 Jul 2021 12:18:45 GMT",
             "content-type": "text/plain",
             "content-length": "11",
             "connection": "close",
             "x-request-id": "01F9XW38B36RSMD4PJ928TK5TB",
             "x-runtime": "0.001247",
             "referrer-policy": "strict-origin-when-cross-origin",
             "ratelimit-observed": "10",
             "ratelimit-remaining": "1990",
             "ratelimit-reset": "1625573985",
             "ratelimit-resettime": "Tue, 06 Jul 2021 12:19:45 GMT",
             "ratelimit-limit": "2000",
             "gitlab-lb": "fe-05-lb-gprd",
             "gitlab-sv": "localhost",
             "cf-cache-status": "DYNAMIC",
             "expect-ct": "max-age=604800, report-uri=\"https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct\"",
             "strict-transport-security": "max-age=31536000",
             "x-content-type-options": "nosniff",
             "server": "cloudflare",
             "cf-ray": "66a8c7083d0bec62-DFW"
           },
           "httpVersion": "1.1"
         }
       }
```

</details>

**Have you created a minimal reproduction repository?**

Please read the [minimal reproductions documentation](https://github.com/renovatebot/renovate/blob/main/docs/development/minimal-reproductions.md) to learn how to make a good minimal reproduction repository.

- [x] I have provided a minimal reproduction repository
- [ ] I don't have time for that, but it happens in a public repository I have linked to
- [ ] I don't have time for that, and cannot share my private repository
- [ ] The nature of this bug means it's impossible to reproduce publicly

Reproduction available in https://gitlab.com/nfm/renovate-gemfile-lock-bug.

Very basic self-hosted setup, using the https://gitlab.com/renovate-bot/renovate-runner pipeline template as a baseline.

The Gemfile in the repo just contains the following:

```
source 'https://rubygems.org'

ruby "2.7.3"

gem "newrelic_rpm", "~>7.1.0"
```

At the time of writing, there's a version bump of `newrelic_rpm` to `7.2.0` available.

Here's [the logs for the CI job that invoked renovate](https://gitlab.com/nfm/renovate-gemfile-lock-bug/-/jobs/1405734488), which contains the Ruby version mismatch error message. Note that the job fails, but [this MR has been opened to update newrelic_rpm](https://gitlab.com/nfm/renovate-gemfile-lock-bug/-/merge_requests/2).

Within the MR to update newrelic_rpm, only the `Gemfile` has been updated.

As a result, attempting to [bundle install for deployment](https://bundler.io/v2.2/man/bundle-config.1.html) (with `BUNDLE_DEPLOYMENT`) fails on that branch, as the `Gemfile` and `Gemfile.lock` aren't compatible.

**Additional context**

I tried to reproduce this using the GitHub installable app version of Renovate, and didn't hit the same issue.

I'm not really familiar with the internals but given the self-hosted version is breaking, it's probably some difference of approach between those two that's causing this.

I can help debug if more info would be useful :pray:

**Comment by viceice** [comment]
Are you using the dind version? Renovate can only use the right ruby version when using the dind renovate, because then renovate starts the lock file update in seperate docker sidecar container and can use a matching ruby version.

https://hub.docker.com/r/renovate/ruby/tags?page=1&ordering=last_updated&name=2.7.3

**Reply by nfm** [reply]
I see - no, the reproduction repo is not using the `dind` version.
 
It's interesting that there's a coupling there between the ruby version baked into the renovate image, and how it processes the `Gemfile`.

Would it be viable to strip the `ruby 'x.y.z'` from the `Gemfile` if present? I'm not familiar with the renovate internals - is there anything it needs to account when bumping versions that depends on the ruby version?

**Reply by viceice** [reply]
No, renovate won't stip anything from `Gemfile`.

The full image (`binarySource=global`) simply uses the ruby version globally available. 
So you can build your own renovate image for your use case or better use the dind version.

**Reply by nfm** [reply]
> No, renovate won't stip anything from Gemfile.

Oh yeah, I know it's not doing that currently - just wondering about potential solutions.

Is there necessary behaviour from maintaining the `Gemfile` from a particular ruby version, or is it just a quirk of the implementation that they need to match?

**Reply by viceice** [reply]
It's a bundler requirement that ruby version must match the `ruby 'x.y.z'` spec. So renovate tries to use the best matching version (`binarySource=docker`).

**Reply by nfm** [reply]
In case it's helpful to anyone, we ended up resolving this by switching from the docker version of renovate to installing renovate as an NPM package within our docker container and running it there. That way, we were guaranteed to have a matching ruby version with what our `Gemfile` was specifying.

Thanks for pointing me in the right direction @viceice :grin:

**Reply by narwold** [reply]
I'm running into this same issue. We use the same Renovate run for Node (pnpm) + Ruby. To clarify, @viceice, is the `dind` version you're referring to the `renovate/ruby` image you posted above? If so, does this have any issues supporting Node?

**Reply by narwold** [reply]
I found this: https://gitlab.com/renovate-bot/renovate-runner

Is there any instruction on doing this with other CI providers? We use Buildkite, but I can deduce with more examples with, say Github Actions for example.

**Reply by narwold** [reply]
For anyone else that stumbles on my question, this is what I had to do:
- `binarySource: 'docker'` and `dockerUser: '0:0'` in `renovate-config.js`
- Volumes `/var/run/docker.sock:/var/run/docker.sock` and `/tmp:/tmp` mounted in Buildkite's Docker plugin
- `privileged: true` and `propagate-uid-gid: true` in Buildkite's Docker plugin

**Reply by FloraChenCerner** [reply]
I'm having the same issue and we are using github version in our company instance. How can I specify the `dind` version for renovate there? @viceice

**Reply by viceice** [reply]
We have an official GitHub action

https://github.com/renovatebot/github-action
